x/ecocredit/server/core/bridge_receive_test.go

```diff
@@ -0,0 +1,316 @@
+}
+
+func TestBridgeReceive_TooManyBatches(t *testing.T) {
+	t.Parallel()
+	refId := "VCS-001"
+	s := setupBase(t)
+	project, batch := setupBridgeTest(s, refId)
+
+	// create 2 batches with "hi" as metadata.
+	batchMetadata := "hi"
+	start, end := batch.StartDate.AsTime(), batch.EndDate.AsTime()
+	_, err := s.k.CreateBatch(s.ctx, &core.MsgCreateBatch{
+		Issuer:    s.addr.String(),
+		ProjectId: project.Id,
+		Issuance: []*core.BatchIssuance{
+			{Recipient: s.addr.String(), TradableAmount: "10"},
+		},
+		Metadata:  batchMetadata,
+		StartDate: &start,
+		EndDate:   &end,
+		Open:      true,
+		OriginTx:  &core.OriginTx{Id: "0x12345", Source: "polygon:0x12345"},
+		Note:      "hi",
+	})
+	assert.NilError(t, err)
+	_, err = s.k.CreateBatch(s.ctx, &core.MsgCreateBatch{
+		Issuer:    s.addr.String(),
+		ProjectId: project.Id,
+		Issuance: []*core.BatchIssuance{
+			{Recipient: s.addr.String(), TradableAmount: "10"},
+		},
+		Metadata:  batchMetadata,
+		StartDate: &start,
+		EndDate:   &end,
+		Open:      true,
+		OriginTx:  &core.OriginTx{Id: "0x123456", Source: "polygon:0x12345"},
+		Note:      "hi",
+	})
+	assert.NilError(t, err)
+
+	msg := core.MsgBridgeReceive{
+		Issuer: s.addr.String(),
+		Batch: &core.MsgBridgeReceive_Batch{
+			Recipient: testutil.GenAddress(),
+			Amount:    "3",
+			StartDate: &start,
+			EndDate:   &end,
+			Metadata:  batchMetadata,
+			Note:      "bridged",
+		},
+		Project: &core.MsgBridgeReceive_Project{
+			ReferenceId:  refId,
+			Jurisdiction: "US-KY",
+			Metadata:     "hi",
+			ClassId:      "C01",
+		},
+	}
+	_, err = s.k.BridgeReceive(s.ctx, &msg)
+	assert.ErrorIs(t, err, sdkerrors.ErrInvalidRequest.Wrapf("fatal error: bridge service %s has %d batches issued "+
+		"with start %v and end %v dates in project %s", s.addr.String(), 2, msg.Batch.StartDate.String(), msg.Batch.EndDate.String(), project.Id))
+}
+
+func TestBridgeReceive_MultipleProjects(t *testing.T) {
+	t.Parallel()
+	refId := "VCS-001"
+	s := setupBase(t)
+	project, _ := setupBridgeTest(s, refId)
+
+	project2 := &api.Project{
+		Id:           "C01-002",
+		Admin:        s.addr,
+		ClassKey:     project.ClassKey,
+		Jurisdiction: "US-KY",
+		Metadata:     "project2",
+		ReferenceId:  refId,
+	}
+	assert.NilError(t, s.stateStore.ProjectTable().Insert(s.ctx, project2))
+
+	start, end := time.Now(), time.Now()
+	msg := core.MsgBridgeReceive{
+		Issuer: s.addr.String(),
+		Batch: &core.MsgBridgeReceive_Batch{
+			Recipient: s.addr.String(),
+			Amount:    "3",
+			OriginTx: &core.OriginTx{
+				Id:     "0x12345",
+				Source: "polygon",
+			},
+			StartDate: &start,
+			EndDate:   &end,
+		},
+		Project: &core.MsgBridgeReceive_Project{
+			ReferenceId:  refId,
+			Jurisdiction: "US-KY",
+			Metadata:     "",
+			ClassId:      "C01",
+		},
+	}
+	res, err := s.k.BridgeReceive(s.ctx, &msg)
+	assert.NilError(t, err)
+	// check to make sure the first project is selected
+	assert.Equal(t, res.ProjectId, project.Id)
```

Comment:
I know this is an edge case, but I'm curious -- do we actually have guaranteed ordering here?

Reply:
yeah both the [string codec](https://github.com/cosmos/cosmos-sdk/blob/main/orm/encoding/ormfield/string.go#L22-L24) and [uint64 codec](https://github.com/cosmos/cosmos-sdk/blob/main/orm/encoding/ormfield/uint64.go#L22-L24) are ordered fields. although i'm not sure if combining them as index keys makes any difference... maybe @aaronc  can confirm?